File: ts-sdk/src/integrations/fides.ts
/**
 * FIDES (Trusted Agent Protocol) integration for agit.
 *
 * Provides DID-signed commits, trust-gated repository access,
 * and verifiable agent identity for multi-agent swarm versioning.
 *
 * @example
 * ```ts
 * import { AgitFidesClient } from "@agit/sdk";
 * import { Fides } from "@fides/sdk";
 *
 * const fides = new Fides({ discoveryUrl: "...", trustUrl: "..." });
 * const identity = await fides.createIdentity({ name: "my-agent" });
 *
 * const agitFides = new AgitFidesClient(agitClient, fides);
 * await agitFides.signedCommit(state, "checkpoint after tool call");
 * ```
 */

import type { AgitClient } from "../client.js";
import { ActionType } from "../types.js";

/** Fides identity attached to a signed commit. */
export interface FidesCommitIdentity {
  did: string;
  publicKey: string;
  signature: string;
  algorithm: "ed25519";
}

/** A commit that has been signed with a fides DID. */
export interface SignedCommit {
  hash: string;
  did: string;
  signature: string;
  trustLevel?: number;
}

/** Options for trust-gated operations. */
export interface TrustGateOptions {
  /** Minimum trust level (0-100) required for the operation. */
  minTrustLevel: number;
  /** DID of the agent requesting the operation. */
  requesterDid: string;
}

/**
 * AgitFidesClient — agit client with FIDES identity and trust.
 *
 * Every commit is signed with the agent's Ed25519 DID keypair,
 * creating a cryptographically verifiable audit trail where each
 * state change is linked to a proven agent identity.
 */
export class AgitFidesClient {
  private client: AgitClient;
  private fides: any; // Fides instance from @fides/sdk
  private did: string | null = null;

  constructor(client: AgitClient, fides: any) {
    this.client = client;
    this.fides = fides;
  }

  /**
   * Initialize with an existing fides identity or create a new one.
   */
  async init(existingDid?: string): Promise<string> {
    if (existingDid) {
      this.did = existingDid;
    } else {
      const identity = await this.fides.createIdentity({
        name: `agit-agent-${Date.now()}`,
        description: "agit-managed agent with fides identity",
      });
      this.did = identity.did;
    }

    // Commit identity creation as system event
    await this.client.commit({
      memory: {
        fides_identity: {
          did: this.did,
          initialized_at: new Date().toISOString(),
        },
      },
      world_state: { fides_phase: "identity_init" },
      message: `fides-init: ${this.did}`,
      action_type: ActionType.SystemEvent,
    });

    return this.did;
  }

  /**
   * Create a DID-signed commit. The commit message includes the
   * agent's DID and Ed25519 signature over the state hash.
   */
  async signedCommit(
    state: Record<string, unknown>,
    message: string,
    actionType: ActionType = ActionType.Checkpoint
  ): Promise<SignedCommit> {
    if (!this.did) {
      throw new Error("Fides identity not initialized. Call init() first.");
    }

    // Serialize state for signing
    const stateJson = JSON.stringify(state, null, 0);
    const stateRequest = {
      method: "POST",
      url: `agit://commit/${this.did}`,
      headers: {
        "content-type": "application/json",
      } as Record<string, string>,
      body: stateJson,
    };

    // Sign with fides (RFC 9421 HTTP signature)
    const signed = await this.fides.signRequest(stateRequest);
    const signature = signed.headers["Signature"] || signed.headers["signature"] || "";

    // Embed fides identity in the committed state
    const enrichedState = {
      ...state,
      _fides: {
        did: this.did,
        signature,
        algorithm: "ed25519",
        signed_at: new Date().toISOString(),
      } satisfies FidesCommitIdentity & { signed_at: string },
    };

    // Commit with agit
    const commitResult = await this.client.commit({
      memory: enrichedState,
      world_state: state.world_state as Record<string, unknown> | undefined,
      message: `[${this.did.slice(0, 20)}...] ${message}`,
      action_type: actionType,
    });

    return {
      hash: (commitResult as any)?.hash ?? "unknown",
      did: this.did,
      signature,
    };
  }

  /**
   * Verify that a commit was signed by a specific agent DID.
   */
  async verifyCommit(commitHash: string): Promise<{
    valid: boolean;
    did?: string;
    error?: string;
  }> {
    try {
      const state = await this.client.getState(commitHash);
      const fidesData = (state as any)?._fides;

      if (!fidesData?.did || !fidesData?.signature) {
        return { valid: false, error: "Commit has no fides signature" };
      }

      // Resolve the DID to get the public key
      const identity = await this.fides.resolve(fidesData.did);
      if (!identity) {
        return { valid: false, error: `Could not resolve DID: ${fidesData.did}` };
      }

      // Reconstruct the request for verification
      const stateWithoutFides = { ...state };
      delete (stateWithoutFides as any)._fides;

      const request = {
        method: "POST",
        url: `agit://commit/${fidesData.did}`,
        headers: {
          "content-type": "application/json",
          signature: fidesData.signature,
          "signature-input": `sig1=("@method" "@target-uri" "content-type");keyid="${fidesData.did}";alg="ed25519"`,
        },
        body: JSON.stringify(stateWithoutFides, null, 0),
      };

      const result = await this.fides.verifyRequest(request);
      return {
        valid: result.valid,
        did: fidesData.did,
        error: result.error,
      };
    } catch (e) {
      return { valid: false, error: String(e) };
    }
  }

  /**
   * Trust-gated merge: only merge if the source branch's last
   * committer has sufficient trust from the current agent.
   */
  async trustedMerge(
    branch: string,
    options: TrustGateOptions
  ): Promise<{ merged: boolean; reason?: string; trustLevel?: number }> {
    if (!this.did) {
      throw new Error("Fides identity not initialized.");
    }

    // Get the latest commit on the source branch
    const commits = await this.client.log({ branch, limit: 1 });
    if (commits.length === 0) {
      return { merged: false, reason: "Branch has no commits" };
    }

    const state = await this.client.getState(commits[0].hash);
    const committerDid = (state as any)?._fides?.did;

    if (!committerDid) {
      return { merged: false, reason: "Branch head has no fides identity" };
    }

    // Check trust level
    try {
      const reputation = await this.fides.getReputation(committerDid);
      const trustLevel = Math.round(reputation.score * 100);

      if (trustLevel < options.minTrustLevel) {
        // Commit the rejection as audit event
        await this.client.commit({
          memory: {
            trust_gate: {
              action: "merge_rejected",
              branch,
              committer_did: committerDid,
              trust_level: trustLevel,
              required_level: options.minTrustLevel,
            },
          },
          message: `fides-gate: merge rejected (trust=${trustLevel} < required=${options.minTrustLevel})`,
          action_type: ActionType.SystemEvent,
        });

        return {
          merged: false,
          reason: `Insufficient trust: ${trustLevel} < ${options.minTrustLevel}`,
          trustLevel,
        };
      }

      // Trust sufficient — perform merge
      await this.client.merge({ branch, strategy: "three-way" as any });

      // Commit trust verification as audit trail
      await this.client.commit({
        memory: {
          trust_gate: {
            action: "merge_approved",
            branch,
            committer_did: committerDid,
            trust_level: trustLevel,
            required_level: options.minTrustLevel,
          },
        },
        message: `fides-gate: merge approved (trust=${trustLevel} >= required=${options.minTrustLevel})`,
        action_type: ActionType.SystemEvent,
      });

      return { merged: true, trustLevel };
    } catch (e) {
      return { merged: false, reason: `Trust check failed: ${e}` };
    }
  }

  /**
   * Issue a trust attestation to another agent and commit it.
   */
  async trustAgent(
    subjectDid: string,
    level: number
  ): Promise<{ attestation: any; commitHash?: string }> {
    if (!this.did) {
      throw new Error("Fides identity not initialized.");
    }

    const attestation = await this.fides.trust(subjectDid, level);

    // Commit trust attestation as audit trail
    const result = await this.client.commit({
      memory: {
        trust_attestation: {
          issuer: this.did,
          subject: subjectDid,
          level,
          attestation_id: attestation.id,
        },
      },
      world_state: { fides_phase: "trust_attestation" },
      message: `fides-trust: ${this.did.slice(0, 16)}... -> ${subjectDid.slice(0, 16)}... (level=${level})`,
      action_type: ActionType.SystemEvent,
    });

    return { attestation, commitHash: (result as any)?.hash };
  }

  /** Get the current agent's DID. */
  getDid(): string | null {
    return this.did;
  }
}

File: python/agit/integrations/fides.py
"""FIDES (Trusted Agent Protocol) integration for agit.

Provides DID-signed commits, trust-gated repository access,
and verifiable agent identity for multi-agent swarm versioning.

Usage::

    from agit.integrations.fides import AgitFidesEngine
    engine = AgitFidesEngine(
        repo_path="./my_repo",
        agent_id="trusted-agent",
        discovery_url="http://localhost:3000",
        trust_url="http://localhost:3001",
    )
    await engine.init_identity(name="my-agent")
    engine.signed_commit(state, "checkpoint after tool call")
"""
from __future__ import annotations

import hashlib
import json
import logging
import time
from typing import Any

from agit.engine.executor import ExecutionEngine

logger = logging.getLogger("agit.integrations.fides")

try:
    from nacl.signing import SigningKey, VerifyKey  # type: ignore[import]
    from nacl.encoding import HexEncoder, RawEncoder  # type: ignore[import]
    import nacl.utils  # type: ignore[import]

    _NACL_AVAILABLE = True
except ImportError:
    _NACL_AVAILABLE = False

try:
    import httpx  # type: ignore[import]

    _HTTPX_AVAILABLE = True
except ImportError:
    _HTTPX_AVAILABLE = False


class FidesIdentity:
    """Local fides identity with Ed25519 keypair and DID."""

    def __init__(self, signing_key: Any, did: str, public_key_hex: str) -> None:
        self.signing_key = signing_key
        self.did = did
        self.public_key_hex = public_key_hex

    @classmethod
    def generate(cls) -> "FidesIdentity":
        """Generate a new Ed25519 identity."""
        if not _NACL_AVAILABLE:
            raise ImportError(
                "PyNaCl is required for fides integration. "
                "Install with: pip install pynacl"
            )
        import base58  # type: ignore[import]

        signing_key = SigningKey.generate()
        public_key_bytes = signing_key.verify_key.encode()
        did = f"did:fides:{base58.b58encode(public_key_bytes).decode()}"
        public_key_hex = public_key_bytes.hex()
        return cls(signing_key, did, public_key_hex)

    def sign(self, message: bytes) -> str:
        """Sign a message and return hex-encoded signature."""
        signed = self.signing_key.sign(message)
        return signed.signature.hex()

    @staticmethod
    def verify(message: bytes, signature_hex: str, public_key_hex: str) -> bool:
        """Verify an Ed25519 signature."""
        if not _NACL_AVAILABLE:
            return False
        try:
            verify_key = VerifyKey(bytes.fromhex(public_key_hex))
            verify_key.verify(message, bytes.fromhex(signature_hex))
            return True
        except Exception:
            return False


class AgitFidesEngine:
    """ExecutionEngine wrapper with FIDES identity and trust.

    Every commit is signed with the agent's Ed25519 DID keypair,
    creating a cryptographically verifiable audit trail where each
    state change is linked to a proven agent identity.
    """

    def __init__(
        self,
        repo_path: str,
        agent_id: str = "fides-agent",
        *,
        discovery_url: str = "http://localhost:3000",
        trust_url: str = "http://localhost:3001",
    ) -> None:
        self._engine = ExecutionEngine(repo_path, agent_id=agent_id)
        self._discovery_url = discovery_url
        self._trust_url = trust_url
        self._identity: FidesIdentity | None = None

    @property
    def engine(self) -> ExecutionEngine:
        return self._engine

    @property
    def did(self) -> str | None:
        return self._identity.did if self._identity else None

    async def init_identity(
        self,
        *,
        name: str = "agit-agent",
        existing_identity: FidesIdentity | None = None,
    ) -> str:
        """Initialize or load a fides identity and register with discovery."""
        if existing_identity:
            self._identity = existing_identity
        else:
            self._identity = FidesIdentity.generate()

        # Register with discovery service
        await self._register_with_discovery(name)

        # Commit identity init as system event
        try:
            self._engine.commit_state(
                {
                    "memory": {
                        "fides_identity": {
                            "did": self._identity.did,
                            "public_key": self._identity.public_key_hex,
                            "initialized_at": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
                        },
                    },
                    "world_state": {"fides_phase": "identity_init"},
                },
                message=f"fides-init: {self._identity.did}",
                action_type="system_event",
            )
        except Exception:
            logger.warning("Failed to commit identity init", exc_info=True)

        return self._identity.did

    def signed_commit(
        self,
        state: dict[str, Any],
        message: str,
        action_type: str = "checkpoint",
    ) -> str | None:
        """Create a DID-signed commit.

        The state is enriched with a _fides field containing the
        agent's DID and Ed25519 signature over the state hash.
        """
        if not self._identity:
            raise RuntimeError("Fides identity not initialized. Call init_identity() first.")

        # Hash the state for signing
        state_json = json.dumps(state, sort_keys=True, default=str)
        state_hash = hashlib.sha256(state_json.encode()).hexdigest()

        # Sign the hash with Ed25519
        signature = self._identity.sign(state_hash.encode())

        # Enrich state with fides identity proof
        enriched_state = {
            **state,
            "_fides": {
                "did": self._identity.did,
                "public_key": self._identity.public_key_hex,
                "signature": signature,
                "state_hash": state_hash,
                "algorithm": "ed25519",
                "signed_at": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            },
        }

        try:
            return self._engine.commit_state(
                enriched_state,
                message=f"[{self._identity.did[:20]}...] {message}",
                action_type=action_type,
            )
        except Exception:
            logger.warning("Failed to create signed commit", exc_info=True)
            return None

    def verify_commit(self, commit_hash: str) -> dict[str, Any]:
        """Verify that a commit was signed by a valid fides identity.

        Returns dict with 'valid', 'did', and optionally 'error'.
        """
        try:
            state = self._engine.get_state_at(commit_hash)
            if not state:
                return {"valid": False, "error": "Commit not found"}

            fides_data = state.get("_fides")
            if not fides_data:
                return {"valid": False, "error": "Commit has no fides signature"}

            did = fides_data.get("did", "")
            signature = fides_data.get("signature", "")
            public_key = fides_data.get("public_key", "")
            stored_hash = fides_data.get("state_hash", "")

            if not all([did, signature, public_key, stored_hash]):
                return {"valid": False, "error": "Incomplete fides data"}

            # Reconstruct state without _fides for hash verification
            state_without_fides = {k: v for k, v in state.items() if k != "_fides"}
            state_json = json.dumps(state_without_fides, sort_keys=True, default=str)
            computed_hash = hashlib.sha256(state_json.encode()).hexdigest()

            # Verify hash matches
            if computed_hash != stored_hash:
                return {"valid": False, "did": did, "error": "State hash mismatch (tampering detected)"}

            # Verify Ed25519 signature
            valid = FidesIdentity.verify(stored_hash.encode(), signature, public_key)
            if not valid:
                return {"valid": False, "did": did, "error": "Ed25519 signature verification failed"}

            return {"valid": True, "did": did}

        except Exception as e:
            return {"valid": False, "error": str(e)}

    async def trusted_merge(
        self,
        branch: str,
        *,
        min_trust_level: int = 50,
    ) -> dict[str, Any]:
        """Trust-gated merge: only merge if the source branch's committer
        has sufficient trust from the current agent.
        """
        if not self._identity:
            raise RuntimeError("Fides identity not initialized.")

        # Get the latest commit on the source branch
        try:
            history = self._engine.get_history(limit=1)
        except Exception:
            return {"merged": False, "reason": "Could not read branch history"}

        if not history:
            return {"merged": False, "reason": "Branch has no commits"}

        latest = history[0]
        committer_did = None

        try:
            state = self._engine.get_state_at(latest.get("hash", ""))
            committer_did = (state or {}).get("_fides", {}).get("did")
        except Exception:
            pass

        if not committer_did:
            return {"merged": False, "reason": "Branch head has no fides identity"}

        # Query trust score
        trust_level = await self._get_trust_score(committer_did)

        if trust_level < min_trust_level:
            self._engine.commit_state(
                {
                    "memory": {
                        "trust_gate": {
                            "action": "merge_rejected",
                            "branch": branch,
                            "committer_did": committer_did,
                            "trust_level": trust_level,
                            "required_level": min_trust_level,
                        },
                    },
                },
                message=f"fides-gate: merge rejected (trust={trust_level} < required={min_trust_level})",
                action_type="system_event",
            )
            return {
                "merged": False,
                "reason": f"Insufficient trust: {trust_level} < {min_trust_level}",
                "trust_level": trust_level,
            }

        # Trust sufficient — merge
        try:
            self._engine.merge(branch, strategy="three_way")
        except Exception as e:
            return {"merged": False, "reason": f"Merge failed: {e}"}

        self._engine.commit_state(
            {
                "memory": {
                    "trust_gate": {
                        "action": "merge_approved",
                        "branch": branch,
                        "committer_did": committer_did,
                        "trust_level": trust_level,
                        "required_level": min_trust_level,
                    },
                },
            },
            message=f"fides-gate: merge approved (trust={trust_level} >= required={min_trust_level})",
            action_type="system_event",
        )
        return {"merged": True, "trust_level": trust_level}

    async def trust_agent(self, subject_did: str, level: int) -> dict[str, Any]:
        """Issue a trust attestation and commit it to the audit trail."""
        if not self._identity:
            raise RuntimeError("Fides identity not initialized.")

        attestation: dict[str, Any] = {}

        # Create and submit attestation via HTTP
        if _HTTPX_AVAILABLE:
            try:
                async with httpx.AsyncClient() as client:
                    payload = {
                        "issuerDid": self._identity.did,
                        "subjectDid": subject_did,
                        "trustLevel": level,
                    }
                    # Sign the attestation payload
                    payload_json = json.dumps(payload, sort_keys=True)
                    signature = self._identity.sign(payload_json.encode())
                    payload["signature"] = signature
                    payload["payload"] = payload_json

                    resp = await client.post(
                        f"{self._trust_url}/attestations",
                        json=payload,
                    )
                    if resp.status_code < 300:
                        attestation = resp.json()
            except Exception as e:
                logger.warning("Failed to submit trust attestation: %s", e)
                attestation = {"error": str(e), "level": level}
        else:
            attestation = {"local_only": True, "level": level}

        # Commit to audit trail
        try:
            self._engine.commit_state(
                {
                    "memory": {
                        "trust_attestation": {
                            "issuer": self._identity.did,
                            "subject": subject_did,
                            "level": level,
                            "attestation": attestation,
                        },
                    },
                    "world_state": {"fides_phase": "trust_attestation"},
                },
                message=f"fides-trust: {self._identity.did[:16]}... -> {subject_did[:16]}... (level={level})",
                action_type="system_event",
            )
        except Exception:
            logger.warning("Failed to commit trust attestation", exc_info=True)

        return attestation

    # ------------------------------------------------------------------
    # Private helpers
    # ------------------------------------------------------------------

    async def _register_with_discovery(self, name: str) -> None:
        """Register identity with the fides discovery service."""
        if not self._identity or not _HTTPX_AVAILABLE:
            return

        try:
            async with httpx.AsyncClient() as client:
                await client.post(
                    f"{self._discovery_url}/identities",
                    json={
                        "did": self._identity.did,
                        "publicKey": self._identity.public_key_hex,
                        "metadata": {
                            "name": name,
                            "type": "agit-agent",
                            "endpoints": {
                                "agit": f"agit://{self._identity.did}",
                            },
                        },
                    },
                )
        except Exception:
            logger.debug("Discovery service not available, operating in local mode")

    async def _get_trust_score(self, did: str) -> int:
        """Query the trust graph for a DID's reputation score."""
        if not _HTTPX_AVAILABLE:
            return 0

        try:
            async with httpx.AsyncClient() as client:
                resp = await client.get(f"{self._trust_url}/scores/{did}")
                if resp.status_code == 200:
                    data = resp.json()
                    return int(data.get("score", 0) * 100)
        except Exception:
            logger.debug("Trust service not available for DID: %s", did)

        return 0
